### ISSUE TITLE
StackOverflowError in Node.isCanCancel / FinalizerGroup in 7.5.1

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#21997]
### Expected Behavior
Just one build failure, the real exception.

### Current Behavior
```
2: Task failed with an exception.
-----------
* What went wrong:
java.lang.StackOverflowError (no error message)

* Exception is:
java.lang.StackOverflowError
        at org.gradle.execution.plan.Node.isCanCancel(Node.java:232)
        at org.gradle.execution.plan.FinalizerGroup.isCanCancel(FinalizerGroup.java:155)
        at org.gradle.execution.plan.CompositeNodeGroup.isCanCancel(CompositeNodeGroup.java:101)
        at org.gradle.execution.plan.Node.isCanCancel(Node.java:232)
        ...
```

### Context
Note:
 * Changing `gradle\wrapper\gradle-wrapper.properties` to 7.5 doesn't show the `StackOverflowError`
 * Running only `gradlew preBuild` doesn't show the `StackOverflowError`
 * Might be related: https://github.com/gradle/gradle/issues/21975

### Steps to Reproduce 
Repro:
 * Clone [this](https://github.com/TWiStErRob/repros/tree/master/gradle/failure-finalizer-stackOverflow) project.
 * Run `gradlew assembleDebug`

### Your Environment

Build scan URL: https://gradle.com/s/psogfmuvy6vcy

**Comment by TWiStErRob** [comment]
Whops, might be a dupicate of https://github.com/gradle/gradle/issues/21542

**Comment by jbartok** [comment]
Could you pls. try with Gradle 7.6 nightly: https://gradle.org/release-nightly/

**Comment by TWiStErRob** [comment]
Sorry, forgot about this.
 * Confirmed fixed in `7.6-20220916231956+0000`
 * Confirmed fixed in `7.6-20220913234616+0000`
 * Confirmed fixed in `7.6-20220822144040+0000`
 * Confirmed fixed in `7.6-20220820035739+0000`
 * Confirmed broken in `7.6-20220819003659+0000`

This confirms it's a duplicate of #21542 and fixed by #21581 (which was merged on 19th August).
